### PR TITLE
Android fixes/improvements

### PR DESCRIPTION
**Pull request by skottmckay**
- Ensure we log an error message before throwing on Android
  - message in exception will be lost due to how the shared libraries are built (both onnxruntime and extensions use static libc++ so there are no shared exception types between them)
- support static or dynamic build of curl/openssl on android
  - TBD which we want to use. static is smaller but build fails on x86 currently so might need a mix
- add infra for anything deriving from BaseKernel to log messages using the ORT logger
  - ensures messages from custom kernels end up in the same place as messages from ORT

### CODE REVIEW

File: operators/azure/logger.hpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  const OrtApi& api_;
+  int api_version_{0};  // api version in use
+  // OrtLogger if available. Unused otherwise
+  const void* ort_logger_{nullptr};
+};
+
+// specialization for when ORT logger is available (compiled against and running with ORT version > 1.15
+// return 'bool'. the actual type doesn't matter - just has to be different to the default of 'void'
+template <>
+Logger<has_Logger_LogMessage<OrtApi>::type>::Logger(const OrtApi& api,
```
Comment (edgchen1) [inline operators/azure/logger.hpp:51]: would the specialization need to be for `std::true_type`? i.e., to ensure that `has_Logger_LogMessage<OrtApi>::value` is actually true.

alternatively, `if constexpr (has_Logger_LogMessage<OrtApi>::value) { ... }` might be a simpler approach that can avoid the template specializations.

Comment (skottmckay) [reply operators/azure/logger.hpp:51]: OrtLogger doesn't exist until 1.15 so I need this to potentially be thrown away due to SFINAE, so constexpr isn't an option (AFAIK).

**Comment by skottmckay** [comment]
> Previously, onnx python windows packages were built with static vc runtime. Then we found it could lead to mysterious hard to debug crashes: [onnx/onnx#3493](https://github.com/onnx/onnx/issues/3493) . It all because they didn't catch all exceptions at binary boundaries.

For now unfortunately there's no other option. The custom op interface for Compute returns void - so there's no way to indicate failure if you do not throw. There are some v2 custom op interface functions recently added to allow Compute to return a status but we'd need to convert all the custom ops.

File: operators/azure/logger.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+// Disable GCC 'ignoring attributes on template argument' warning due to Logger_LogMessage using
+// `__attribute__((warn_unused_result))`.
+// The template here is about whether Logger_LogMessage exists or not, so the attribute is irrelevant.
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wignored-attributes"
+#endif
+
+template <class, class = void>
+struct has_Logger_LogMessage : std::false_type {};
```
Comment (edgchen1) [inline operators/azure/logger.hpp:21]: maybe move `has_Logger_LogMessage` and `LoggerImpl` into an internal `detail` namespace?

File: operators/azure/azure_triton_invoker.cc [hunk 3]
```diff
@@ -98,7 +98,7 @@ int8_t* CreateNonStrTensor(const std::string& data_type,
 
 #define CHECK_TRITON_ERR(ret, msg)                                                    \
   if (!ret.IsOk()) {                                                                  \
-    return ORTX_CXX_API_THROW("Triton err: " + ret.Message(), ORT_RUNTIME_EXCEPTION); \
+    ORTX_CXX_API_THROW("Triton err: " + ret.Message(), ORT_RUNTIME_EXCEPTION); \
```
Comment (edgchen1) [inline operators/azure/azure_triton_invoker.cc:101]: nit: spacing of trailing '\'. also noticed that the `msg` parameter is not used

File: includes/exceptions.h [hunk 4]
```diff
@@ -29,21 +35,23 @@ struct Exception : std::exception {
...
+// Usages:
+//  - logging exception message when they may not propagate up
+//  - logging failure when using the ORT logger
+inline void LogError(const ORTCHAR_T* file, int line, const char* msg) {
 #if defined(__ANDROID__)
-  __android_log_print(ANDROID_LOG_ERROR, "onnxruntime-extensions", "Exception in %s line %d: %s", file, line, msg);
+  __android_log_print(ANDROID_LOG_ERROR, "onnxruntime-extensions", "Error in %s line %d: %s", file, line, msg);
 #else
-  std::cerr << "Exception in " << file << " line " << line << ": " << msg << std::endl;
+  std::cerr << "Error in " << file << " line " << line << ": " << msg << std::endl;
```
Comment (edgchen1) [inline includes/exceptions.h:46]: may need to convert from `const wchar_t*` to `const char*` first if ORTCHAR_T is wchar_t

Comment (edgchen1) [reply includes/exceptions.h:49]: huh, didn't know that `std::wostream` can work with `char*` too

**Review by edgchen1 — APPROVED** [review]

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**